ENH: Give EpochsTFR __getitem__ and __len__ functionality

kdoelling1919 opened the pull request:
#### Reference issue
Addresses some of the points brought up in #5711 


#### What does this implement/fix?
This allows for the selection of a subset of epochs either using slice, indexing or event_id. 


#### Additional information
I had to move `_hid_match` function, which helps with extracting epochs based on event_id, from epochs.py to utils.py in order to call it from tfr.py and avoid rewriting the function there. 

I didn't have time to implement metadata functionality for EpochsTFR. Perhaps I can get to that over the next week. Thought it would be good to start the pull request with this though.

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -2024,8 +2036,94 @@ def __abs__(self):
+        return self._getitem(item)
+
+    def _getitem(self, item, copy=True, drop_event_id=True,
+                 return_indices=False):
```
can you factorize this function with the Epochs class eg with a Mixin object from which both Epochs and EpochsTFR would inheritate?

kdoelling1919 commented:
Thanks @agramfort 

Ok! Factorized so that epochs and epochsTFR both take a GetEpochsMixIn which also does the metadata. I couldn't think of a reason why the get items and metadata should be separated but can do that if you think it better. Good news is metadata functionality is now added.

I think I have adjusted all of the parameters so that they are after the old ones. Will make more changes soon!

kdoelling1919 commented on mne/time_frequency/tfr.py:
```diff
@@ -1990,8 +2000,8 @@ class EpochsTFR(_BaseTFR):
-    def __init__(self, info, data, times, freqs, comment=None,
-                 method=None, verbose=None):  # noqa: D102
+    def __init__(self, info, data, times, freqs, events=None,
+                 event_id=None, comment=None, method=None, verbose=None):  # noqa: D102
```
Does this apply to `verbose` as well? I tend to notice it always being last.

jona-sassenhagen commented:
V. nice @kdoelling1919 

ping @palday

jona-sassenhagen commented on mne/utils.py:
```diff
@@ -3172,3 +3172,354 @@ def _check_if_nan(data, msg=" to be plotted"):
+        `Epochs` or tuple(Epochs, np.ndarray) if `return_indices` is True
+
+        object with subset of epochs (and optionally array with kept
+        epoch indices)
```
This is a bit mangled isn't it?

jona-sassenhagen commented on mne/utils.py:
```diff
@@ -3172,3 +3172,354 @@ def _check_if_nan(data, msg=" to be plotted"):
+        data = self._data
+        del self._data
+        inst = self.copy() if copy else self
+        self._data, inst._data = data, data
```
`self._data = inst._data = data`?

kdoelling1919 replied:
This and a few of the other things you noted are older code that I moved around so that it could be used by both Epochs and EpochsTFR. I don't like to change things too much, in case I am missing why they needed to be this way. But this is simple enough, thanks for pointing it out.

jona-sassenhagen replied:
Yeah I saw eventually that code was just moved around ... but still, if it's cleaner afterwards, even better!

jona-sassenhagen commented on mne/utils.py:
```diff
@@ -3172,3 +3172,354 @@ def _check_if_nan(data, msg=" to be plotted"):
+            pd = _check_pandas_installed(strict=False)
+            if pd is not False:
+                if not isinstance(metadata, pd.DataFrame):
+                    raise TypeError('metadata must be a pandas DataFrame, '
```
use `mne.utils._validate_types`

jona-sassenhagen commented on mne/utils.py:
```diff
@@ -3172,3 +3172,354 @@ def _check_if_nan(data, msg=" to be plotted"):
+            else:
+                if not isinstance(metadata, list):
+                    raise TypeError('metdata must be a list, got %s'
+                                    % (type(metadata),))
```
Also here, etc

jona-sassenhagen commented on mne/utils.py:
```diff
@@ -3172,3 +3172,354 @@ def _check_if_nan(data, msg=" to be plotted"):
+
+    def _keys_to_idx(self, keys):
+        """Find entries in event dict."""
+        keys = [keys] if not isinstance(keys, (list, tuple)) else keys
```
`keys = keys if isinstance(keys, (list, tuple)) else [keys]` right?

kdoelling1919 replied:
These are ultimately the same, right? Happy to change it, but I don't see the advantage of one over the other.

jona-sassenhagen replied:
Not having the not is simpler.

jona-sassenhagen commented:
I love it! Do we need more changes to the docs/website?

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -1960,38 +1962,46 @@ class EpochsTFR(_BaseTFR):
         If not None, override default verbose level (see :func:`mne.verbose`
         and :ref:`Logging documentation <tut_logging>` for more).
+    events : ndarray, shape (n_events, 3)
+        The events as stored in the Epochs class
```
say it can be None and idem for event_id

agramfort commented on mne/time_frequency/tests/test_tfr.py:
```diff
@@ -639,5 +640,46 @@ def test_compute_tfr():
+    time_bandwidth = 4.0  # With 0.5 s time windows, this gives 8 Hz smoothing
+    power = tfr_multitaper(epochs, freqs=freqs, n_cycles=n_cycles,
+                          use_fft=True, time_bandwidth=time_bandwidth,
+                          return_itc=False, average=False, n_jobs=1)
```
you should test all the methods from the mixin class such as __len__ and the iteration with __iter__ and next

agramfort commented:
it would be ideal to illustrate this feature in one of our time frequency example using tfr_epochs 

but I am not sure we have a good dataset for this.

@kdoelling1919 can you go over the examples and if you see any opportunity for documenting this and improving the state of the doc?

thanks a lot for this great new feature

jona-sassenhagen commented:
We can just use aud vis/left right data and demonstrate HID tagging.

agramfort commented:
It does really contain induced effects

jona-sassenhagen commented on mne/time_frequency/tests/test_tfr.py:
```diff
@@ -680,6 +680,23 @@ def test_getitem_epochsTFR():
+        assert_array_equal(power_ep, power.data[ind])
+        ind += 1
+        if ind == 5:
+            break
```
`for ii, power_ep in enumerate(power):`

or

`for ii, power in zip(range(5), power):` (to save the break line)

palday commented:
I just got back from a long trip and it's the busiest time of the year, but I'll take a look when I get a chance. Glad to see this behaviour finally being implemented!

kdoelling1919 commented on mne/utils.py:
```diff
@@ -3172,3 +3173,350 @@ def _check_if_nan(data, msg=" to be plotted"):
+    for key in keys:
+        if not isinstance(key, str):
+            raise KeyError('keys must be strings, got %s (%s)'
+                           % (type(key), key))
```
I tried this change but _validate_types generates a TypeError where a KeyError is what is expected in some tests. So I think reverting back tot he old code makes the most sense. What do you think @jona-sassenhagen ?

kdoelling1919 commented:
OK, I believe all requested changes have been made and the pull request is ready to go! The only snag is I'm not sure how to fix the error in The Travis CI build. It seems to be due to a timeout in `test_scale_mri_xfm` but I can't see for the life of me what that has to do with the changes in this pull request. Any help with this would be greatly appreciated!

Would also appreciate feedback on the notes I left in this [comment](https://github.com/mne-tools/mne-python/pull/5762#issuecomment-443816357)!

larsoner commented:
`test_scale_mri_xfm` has been failing in `master` on the OSX build, you can ignore this one (I'll fix it separately)

These are real, though

https://travis-ci.org/mne-tools/mne-python/jobs/466239479#L3297

I'll let other comment on the substantive questions in the meantime

kdoelling1919 commented:
> `test_scale_mri_xfm` has been failing in `master` on the OSX build, you can ignore this one (I'll fix it separately)
> 
> These are real, though
> 
> https://travis-ci.org/mne-tools/mne-python/jobs/466239479#L3297
> 
> I'll let other comment on the substantive questions in the meantime

Great. Thanks for pointing out the lines. I was having trouble finding the error there.

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -630,7 +631,17 @@ def _tfr_aux(method, inst, freqs, decim, return_itc, picks, average,
+            meta = inst._metadata.copy() if inst._metadata is not None \
+                else None
+            evs = inst.events.copy()
+            ev_id = inst.event_id.copy()
```
use

copy.deepcopy(inst.event_id)

as event_id can be just an integer or a list of integer for some epochs.

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -1960,38 +1971,46 @@ class EpochsTFR(_BaseTFR):
-    def __init__(self, info, data, times, freqs, comment=None,
-                 method=None, verbose=None):  # noqa: D102
+    def __init__(self, info, data, times, freqs, comment=None, method=None,
+                 verbose=None,  events=None, event_id=None, metadata=None):
```
verbose is the exception to what I said in my last review. verbose is always the last parameter.

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -2022,21 +2044,9 @@ def __repr__(self):  # noqa: D105
-        return EpochsTFR(info=self.info.copy(), data=self.data.copy(),
-                         times=self.times.copy(), freqs=self.freqs.copy(),
-                         method=self.method, comment=self.comment)
+        epochs = self.copy()
```
where is the copy() method implemented now?

kdoelling1919 replied:
In the class _BaseTFR, mne/time_frequency/tfr.py:910 (line 899, in master).

agramfort commented on mne/time_frequency/tests/test_tfr.py:
```diff
@@ -640,4 +642,64 @@ def test_compute_tfr():
+    n_events = 10
+
+    # create fake metadata
+    rt = np.random.uniform(size=(n_events,))
```
we prefer for fix everywhere the RandomState to avoid random bugs in tests. I know it should not matter here but I prefer we stick to this rule. Please use :

rng = np.random.RandomState(42)
rt = rng.uniform(...)

and remove the calls to np.random below.

thanks

agramfort commented:
ok thanks. nice

kdoelling1919 commented:
The circleci test seems to have failed for reasons unrelated to the PR. Is that right? Or is it something I can fix?

kdoelling1919 commented:
The rebase appears to have been successful! Modulo the bottom half of [this comment](https://github.com/mne-tools/mne-python/pull/5762#issuecomment-443816357), I think we are ready for merge. Any other comments?

agramfort commented:
cool

@kdoelling1919 can you update whats_new.rst?

I would suggest to the second half in a subsequent PR.

agramfort commented:
+1 for MRG when CIs are green

kdoelling1919 commented:
Argh, so close! I'm quite sure the [error](https://travis-ci.org/mne-tools/mne-python/jobs/467695114#L2631) in Travis isn't related to this PR. `test_fetch_parcellations` runs smoothly on my computer, as does `make test`

larsoner commented:
Unrelated timeout, restarted

kdoelling1919 commented:
Thanks for comments @agramfort @larsoner @jona-sassenhagen 

That was very satisfying

jona-sassenhagen commented:
Thanks for implementing this Keith. I'd been waiting for a while.
